include/pwm.h

```diff
@@ -0,0 +1,138 @@
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#ifndef PWM_H_
+#define PWM_H_
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include <stdint.h>
+#include <stdbool.h>
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+/**
+ * @enum pwm_polarity
+ * @brief Possible polarities of the PWM signal
+ */
+enum pwm_polarity {
+	PWM_POLARITY_NORMAL,
+	PWM_POLARITY_INVERSED,
+};
+
+/**
+ * @struct pwm_init_param
+ * @brief  Structure containing the init parameters needed by the PWM generator
+ */
+struct pwm_init_param {
+	/** Base address where the PWM generator core is situated */
+	uint32_t base_addr;
+	/** PWM generator reference clock rate , in Hz */
+	uint32_t clk_rate_hz;
+	/** PWM generator period */
+	uint32_t period_ns;
+	/** PWM generator duty cycle */
+	uint32_t duty_cycle_ns;
+	/** PWM generator polarity */
+	enum pwm_polarity polarity;
+	/** PWM generator enabled */
+	bool enabled;
+};
+
+/**
+ * @struct pwm_desc
+ * @brief  Structure representing an PWM generator device
+ */
+struct pwm_desc {
+	/** Base address where the PWM generator core is situated */
+	uint32_t base_addr;
+	/** PWM generator reference clock rate, in Hz */
+	uint32_t clk_rate_hz;
+	/** PWM generator period */
+	uint32_t period_ns;
+	/** PWM generator duty cycle */
+	uint32_t duty_cycle_ns;
+	/** PWM generator polarity */
+	enum pwm_polarity polarity;
+	/** PWM generator enabled */
+	bool enabled;
+};
+
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+/* Initialize the PWM generator device */
+int32_t pwm_init(struct pwm_desc **desc,
+		 const struct pwm_init_param *param);
+
+/* Free the resources used by the PWM generator device */
+int32_t pwm_remove(struct pwm_desc *desc);
+
+/* Enable PWM generator device */
+int32_t pwm_enable(struct pwm_desc *desc);
+
+/* Disable PWM generator device */
+int32_t pwm_disable(struct pwm_desc *desc);
+
+/* Set period of PWM generator device */
+int32_t pwm_set_period(struct pwm_desc *desc,
```

Comment:
Just a curiosity on my side: why did you prefer `period` over `frequency`?

I believe the frequency is a more common term when using PWMs.

Reply:
The unit preferred in linux is period. Since duty_cycle is in ns, it makes sense to have PWM period, probably.

Reply:
I'am more familiar with frequency too and with duty expressed in percents, But our direction is to follow how it is done in linux is ok to me